adafruit_ble/scanner.py

```diff
@@ -0,0 +1,188 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2019 Dan Halbert for Adafruit Industries
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_ble.scanner`
+====================================================
+
+Scan for nearby BLE Devices
+
+* Author(s): Dan Halbert for Adafruit Industries
+
+"""
+import struct
+
+import bleio
+
+from .advertising import AdvertisingPacket
+
+class Scanner:
+    """
+    Scan for BLE device advertisements for a period of time. Return what was received.
+
+    Example::
+
+        from adafruit_ble.scanner import Scanner
+        scanner = Scanner()
+        scanner.Scan
+
+        # Wait for a connection.
+        while not uart.connected:
+            pass
+
+        uart.write('abc')
+    """
+
+    def __init__(self):
+        self._scanner = bleio.Scanner()
+
+    def scan_unique(self, timeout, *, interval=0.1, window=0.1):
+        """Scan for advertisements from BLE devices. Suppress duplicates
+        in returned `ScanEntry` objects.
+
+        :param int timeout: how long to scan for (in seconds)
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.
+           Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+          `window` must be <= `interval`.
+        :returns a list of `adafruit_ble.ScanEntry` objects.
+
+        """
+        return ScanEntry.unique(self.scan(timeout, interval=interval, window=window))
```

Comment:
Did you try using `set(self.scan(...))` to achieve this?

Reply:
I considered this, but the trouble is that `set()` requires the items be hashable and implement  `__eq__()`. In this case, we want to ignore the `rssi` value, but that would make `__eq__()` say two things were equal even if their rssi was different, which isn't what we want.

Reply:
Ok, I think this should be renamed `scan_devices` to make it clear what makes them unique. I'd still return a set (frozenset is only in some builds right?) but deduplicate it yourself.

Reply:
I like the method name. Thanks!

It doesn't really help to store it in a `set()` because `set()` is not going to do deduplication because `__eq__()` is not implemented, so it just considers whether they are different objects are not. I could return a tuple.